File: Solutions/543-DiameterOfBinaryTree/DiameterOfBinaryTree.ts
import { TreeNode } from "../../Types";

export function diameterOfBinaryTree(root: TreeNode | null): number {
  let maxDiameter = 0;

  function dfs(node: TreeNode | null) {
    if (!node) return 0;

    const maxHeightLeft = dfs(node.left);
    const maxHeightRight = dfs(node.right);
    const maxHeight = Math.max(maxHeightLeft, maxHeightRight);
    const diameter = maxHeightLeft + maxHeightRight;

    maxDiameter = Math.max(maxDiameter, diameter);
    return maxHeight + 1;
  }

  dfs(root);

  return maxDiameter;
}

File: Solutions/All/543-DiameterOfBinaryTree/DiameterOfBinaryTree.kt
class Solution {
    var maxDiameter: Int = 0

    fun dfs(node: TreeNode?): Int {
        if (node == null) return 0

        val left = dfs(node?.left)
        val right = dfs(node?.right)
        val currentDiameter = left + right

        maxDiameter = Math.max(maxDiameter, currentDiameter)
        return Math.max(left, right) + 1
    }

    fun diameterOfBinaryTree(root: TreeNode?): Int {
        dfs(root)
        return maxDiameter
    }
}
